File: 01-Programming-Languages/Ruby/Data-Structures/Tree/binary_tree_traversal.rb
=begin
Problem: Binary Tree Traversals
Description: Perform inorder, preorder, and postorder traversals on a binary tree.

Approach:
- Inorder: left -> root -> right
- Preorder: root -> left -> right
- Postorder: left -> right -> root

Time Complexity: O(n)
Space Complexity: O(n)

Example:
Input:  Tree: 1 -> (2, 3) where 2 -> (4, 5)
Output:
  Inorder:   [4, 2, 5, 1, 3]
  Preorder:  [1, 2, 4, 5, 3]
  Postorder: [4, 5, 2, 3, 1]
=end

class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def inorder(root, result = [])
  return result if root.nil?
  inorder(root.left, result)
  result << root.val
  inorder(root.right, result)
  result
end

def preorder(root, result = [])
  return result if root.nil?
  result << root.val
  preorder(root.left, result)
  preorder(root.right, result)
  result
end

def postorder(root, result = [])
  return result if root.nil?
  postorder(root.left, result)
  postorder(root.right, result)
  result << root.val
  result
end

root = TreeNode.new(1,
  TreeNode.new(2, TreeNode.new(4), TreeNode.new(5)),
  TreeNode.new(3)
)

puts "Inorder:   #{inorder(root).join(' ')}"
puts "Preorder:  #{preorder(root).join(' ')}"
puts "Postorder: #{postorder(root).join(' ')}"

File: 01-Programming-Languages/JavaScript/Data-Structures/Tree/binaryTreeTraversal.js
/*
Problem: Binary Tree Traversals
Description: Perform inorder, preorder, and postorder traversals on a binary tree recursively.

Approach:
- Inorder: left, root, right
- Preorder: root, left, right
- Postorder: left, right, root

Time Complexity: O(n)
Space Complexity: O(n)

Example:
Input: Tree with root = 1, left = 2, right = 3, left.left = 4, left.right = 5
Output:
  Inorder: [4, 2, 5, 1, 3]
  Preorder: [1, 2, 4, 5, 3]
  Postorder: [4, 5, 2, 3, 1]
*/

class TreeNode {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function inorder(root, result = []) {
  if (!root) return result;
  inorder(root.left, result);
  result.push(root.val);
  inorder(root.right, result);
  return result;
}

function preorder(root, result = []) {
  if (!root) return result;
  result.push(root.val);
  preorder(root.left, result);
  preorder(root.right, result);
  return result;
}

function postorder(root, result = []) {
  if (!root) return result;
  postorder(root.left, result);
  postorder(root.right, result);
  result.push(root.val);
  return result;
}

const root = new TreeNode(1,
  new TreeNode(2, new TreeNode(4), new TreeNode(5)),
  new TreeNode(3)
);

console.log('Inorder:', inorder(root));
console.log('Preorder:', preorder(root));
console.log('Postorder:', postorder(root));
